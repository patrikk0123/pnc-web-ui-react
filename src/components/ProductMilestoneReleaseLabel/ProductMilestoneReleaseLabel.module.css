.milestone-release-label * {
  text-decoration: none !important;
}

.milestone-release-label--release {
  cursor: default;
}

.tooltip-text {
  text-align: left;
}
