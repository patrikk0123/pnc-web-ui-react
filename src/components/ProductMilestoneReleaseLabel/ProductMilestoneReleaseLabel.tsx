import { Button, Tooltip } from '@patternfly/react-core';
import { useNavigate } from 'react-router-dom';

import { ProductMilestone, ProductRelease } from 'pnc-api-types-ts';

import { isProductMilestone, isProductRelease } from 'utils/entityRecognition';
import { createDateTime } from 'utils/utils';

import styles from './ProductMilestoneReleaseLabel.module.css';

interface IProductMilestoneReleaseProp {
  productMilestoneRelease: ProductMilestone | ProductRelease;
  isCurrent: boolean;
  link?: string;
}

/**
 * Show Product Milestone or Release label with specific text and toolips.
 *
 * @param productMilestoneRelease - ProductMilestone or ProductRelease component to be displayed
 * @param isCurrent - If the ProductMilestone is current one
 * @param link - Link to navigate to
 */
export const ProductMilestoneReleaseLabel = ({ productMilestoneRelease, isCurrent, link }: IProductMilestoneReleaseProp) => {
  const navigate = useNavigate();

  let tooltipContent;
  let buttonClassName;
  let labelType;
  if (isProductMilestone(productMilestoneRelease)) {
    const productMilestone = productMilestoneRelease as ProductMilestone;
    labelType = 'milestone';
    tooltipContent = (
      <div className={styles['tooltip-text']}>
        <strong>Start Date: </strong>
        {createDateTime({ date: productMilestone.startingDate!, includeTime: false }) || <>&ndash;</>}
        <br />
        <strong>Planned End Date: </strong>
        {createDateTime({ date: productMilestone.plannedEndDate!, includeTime: false }) || <>&ndash;</>}
        <br />
        <strong>End Date: </strong>
        {createDateTime({ date: productMilestone.endDate!, includeTime: false }) || <>&ndash;</>}
      </div>
    );
    buttonClassName = isCurrent ? `${styles['milestone-label']} ${styles['is-current']}` : `${styles['milestone-label']}`;
  } else if (isProductRelease(productMilestoneRelease)) {
    labelType = 'release';
    const productRelease = productMilestoneRelease as ProductRelease;
    tooltipContent = (
      <div className={styles['toolip-text']}>
        <strong>Release Date: </strong>
        {createDateTime({ date: productRelease.releaseDate!, includeTime: false }) || <>&ndash;</>}
        <br />
        <strong>Support Level: </strong>
        {productRelease.supportLevel}
      </div>
    );
    buttonClassName = `${styles['release-label']}`;
  }
  return (
    <span className={styles.label}>
      <Tooltip removeFindDomNode content={tooltipContent} isContentLeftAligned={true} position="auto">
        <Button
          onClick={link ? () => navigate(link) : undefined}
          isSmall={true}
          className={buttonClassName}
          component={labelType === 'milestone' ? 'a' : 'span'}
        >
          {productMilestoneRelease.version}
        </Button>
      </Tooltip>
    </span>
  );
};
