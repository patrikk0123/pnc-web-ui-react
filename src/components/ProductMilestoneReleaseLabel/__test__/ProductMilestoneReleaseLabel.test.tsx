import { render, screen } from '@testing-library/react';
import { MemoryRouter } from 'react-router';
import { vi } from 'vitest';

import { ProductMilestone, ProductRelease } from 'pnc-api-types-ts';

import { ProductMilestoneReleaseLabel } from 'components/ProductMilestoneReleaseLabel/ProductMilestoneReleaseLabel';

import productMilestoneMock from './data/product-milestones-mock.json';
import productReleaseMock from './data/product-releases-mock.json';

vi.mock('services/keycloakService');
vi.mock('services/webConfigService');

describe('display MilestoneReleaseLabel component', () => {
  test('compare snapshot with previous record', () => {
    let testMilestoneDataList = productMilestoneMock as unknown as ProductMilestone[];
    let testReleaseDataList = productReleaseMock as unknown as ProductRelease[];

    const tree = render(
      <MemoryRouter>
        <ProductMilestoneReleaseLabel
          productMilestoneRelease={testMilestoneDataList[0]}
          isCurrent={false}
        ></ProductMilestoneReleaseLabel>
        <ProductMilestoneReleaseLabel
          productMilestoneRelease={testMilestoneDataList[1]}
          isCurrent={true}
        ></ProductMilestoneReleaseLabel>
        <ProductMilestoneReleaseLabel
          productMilestoneRelease={testReleaseDataList[0]}
          isCurrent={false}
        ></ProductMilestoneReleaseLabel>
      </MemoryRouter>
    );
    expect(tree.container).toMatchSnapshot();
  });
});
