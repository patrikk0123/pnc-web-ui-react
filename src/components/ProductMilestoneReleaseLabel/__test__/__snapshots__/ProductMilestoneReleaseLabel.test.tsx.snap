// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`display MilestoneReleaseLabel component compare snapshot with previous record 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <span
        class="label"
      >
        <div
          style="display: contents;"
        >
          <span
            aria-disabled="false"
            class="pf-v5-c-button pf-m-primary pf-m-small milestone-label"
            data-ouia-component-id="OUIA-Generated-Button-primary-4"
            data-ouia-component-type="PF5/Button"
            data-ouia-safe="true"
          >
            ERP1.2.3
          </span>
        </div>
      </span>
      <span
        class="label"
      >
        <div
          style="display: contents;"
        >
          <span
            aria-disabled="false"
            class="pf-v5-c-button pf-m-primary pf-m-small milestone-label is-current"
            data-ouia-component-id="OUIA-Generated-Button-primary-5"
            data-ouia-component-type="PF5/Button"
            data-ouia-safe="true"
          >
            ERP1.2.4-C
          </span>
        </div>
      </span>
      <span
        class="label"
      >
        <div
          style="display: contents;"
        >
          <span
            aria-disabled="false"
            class="pf-v5-c-button pf-m-primary pf-m-small release-label"
            data-ouia-component-id="OUIA-Generated-Button-primary-6"
            data-ouia-component-type="PF5/Button"
            data-ouia-safe="true"
          >
            CAM1.2.5
          </span>
        </div>
      </span>
    </div>
  </body>,
  "container": <div>
    <span
      class="label"
    >
      <div
        style="display: contents;"
      >
        <span
          aria-disabled="false"
          class="pf-v5-c-button pf-m-primary pf-m-small milestone-label"
          data-ouia-component-id="OUIA-Generated-Button-primary-4"
          data-ouia-component-type="PF5/Button"
          data-ouia-safe="true"
        >
          ERP1.2.3
        </span>
      </div>
    </span>
    <span
      class="label"
    >
      <div
        style="display: contents;"
      >
        <span
          aria-disabled="false"
          class="pf-v5-c-button pf-m-primary pf-m-small milestone-label is-current"
          data-ouia-component-id="OUIA-Generated-Button-primary-5"
          data-ouia-component-type="PF5/Button"
          data-ouia-safe="true"
        >
          ERP1.2.4-C
        </span>
      </div>
    </span>
    <span
      class="label"
    >
      <div
        style="display: contents;"
      >
        <span
          aria-disabled="false"
          class="pf-v5-c-button pf-m-primary pf-m-small release-label"
          data-ouia-component-id="OUIA-Generated-Button-primary-6"
          data-ouia-component-type="PF5/Button"
          data-ouia-safe="true"
        >
          CAM1.2.5
        </span>
      </div>
    </span>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
